### ISSUE TITLE
Inconsistent type inference for Enum() with string literal vs variable name

### ISSUE THREAD

**Issue by Deshdeepak1** [microsoft/pyright#10873]
**Describe the bug**
When creating an Enum using the functional API, Pyright infers different types depending on whether the name is a string literal or a variable.

- If the name is a literal, the inferred type is type[<EnumClass>].
- If the name is a variable (even if it’s just a str constant), the inferred type is Enum

I would expect at least type[Enum] in the second case, otherwise the error propagates to my use cases. e.g.

**Code or Screenshots**

```python
from enum import Enum
from typing import reveal_type

A = Enum("A", {"a": 1})
reveal_type(A)

a = A["a"]

name1 = "B"
B = Enum(name1, {"b": 2})
reveal_type(B)

b = B["b"]

name2 = "C"
C: type[Enum] = Enum(name2, {"c": 3})
reveal_type(C)

c = C["c"]
```

```bash
[deshdeepak@archlinux temp]$ pyright b.py
/home/deshdeepak/temp/b.py
  /home/deshdeepak/temp/b.py:5:13 - information: Type of "A" is "type[A]"
  /home/deshdeepak/temp/b.py:11:13 - information: Type of "B" is "Enum"
  /home/deshdeepak/temp/b.py:13:5 - error: "__getitem__" method not defined on type "Enum" (reportIndexIssue)
  /home/deshdeepak/temp/b.py:16:17 - error: Type "Enum" is not assignable to declared type "type[Enum]"
    Type "Enum" is not assignable to type "type[Enum]" (reportAssignmentType)
  /home/deshdeepak/temp/b.py:17:13 - information: Type of "C" is "type[Enum]"
2 errors, 0 warnings, 3 informations
```

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?
Using as command-line tool as well as a language server in neovim.

```bash
[deshdeepak@archlinux temp]$ pyright --version
pyright 1.1.404
```

**Comment by erictraut** [comment]
Static type checkers like pyright and mypy support the functional form of `Enum` class definitions, but they require that you provide the name (the first argument) and the member definitions (the second argument) as literal expressions. If you use dynamic expressions (like in your `B` and `C` examples), then the special-case logic in the type checker cannot reason about the class. It then falls back on the generic (non-special-cased) handling of constructor calls. The resulting class won't have any of the non-standard behaviors that an Enum normally has.

Not surprisingly, pyright's behavior here matches that of mypy. 

If you want your code to work with static type checkers, I recommend avoiding such dynamic behaviors.

**Comment by Deshdeepak1** [comment]
The behavior doesn't match mypy.
Although mypy gives error on the argument not being string literal,  it's reveal type is accurate.

```
[deshdeepak@archlinux temp]$ mypy b.py 
b.py:5: note: Revealed type is "def (value: builtins.object) -> b.A"
b.py:10: error: Enum() expects a string literal as the first argument  [misc]
b.py:11: note: Revealed type is "def (value: builtins.object) -> b.B"
b.py:16: error: Enum() expects a string literal as the first argument  [misc]
b.py:17: note: Revealed type is "def (value: builtins.object) -> b.C"
Found 2 errors in 1 file (checked 1 source file)
```


>  It then falls back on the generic (non-special-cased) handling of constructor calls. 

Yes, then at least the generic behavior should result in type[Enum] instead of Enum, which is wrong .
